design/one-pager-host-aware-stack-manager.md

```diff
@@ -0,0 +1,211 @@
+ Cluster.
+
+#### Stack Installation
+
+In case of a stack installation, Stack Manager should:
+
+1. Create install job on Host Cluster and read logs of install pod from Host.
+2. Deploy CRDs and RBAC on Tenant Kubernetes
+3. Wait until a service account token secret is created for the stack to be installed.
+4. Copy service account token secret into Host Cluster.
+5. Deploy stack controller deployment by [configuring against Tenant Kubernetes Apiserver](#configuring-a-pod-against-tenant-kubernetes-api-server).
+
+
+##### Mapping Resources to Host
+
+While creating resources (token secrets or controller deployments) on Host Cluster, Stack Manager will map objects
+ (jobs, deployments, secrets ...) on multiple namespaces of Tenant Kubernetes into a single namespace on Host Cluster
+ by prepending `<namespace>.` to object name. For example:
+
+ | Tenant Kubernetes | Host Cluster |
+ | :---: | :---: |
+ | Namespace / Name | Namespace / Name |
+ | crossplane-system / stack-gcp | tenant-n-controllers / crossplane-system.stack-gcp |
+ | dev / stack-wordpress | tenant-n-controllers / dev.stack-wordpress |
+
+This mapping will cause a length limit to resource names as follows:
+
+```
+len(namespace) + len(resource-name) + 1 <= 253
+```
+
+Right now, this will effect:
+- Install job name
+- Stack controller deployment name
+- Stack controller service account name
+
+However, considering the following, this should not cause a new limitation for Crossplane:
+
+- There are already labels holding some resource names (e.g. `core.crossplane.io/parent-name: stack-gcp`) and Kubernetes
+ Labels are [limited to 63 chars](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set)
+- Kubernetes namespaces are limited to 63 chars
+ 
+#### Stack Uninstallation
+
+Stack manager can no longer rely on owner references for deletion of stack install jobs and controller deployments
+ since stack related custom resources and jobs/deployments live in different Kubernetes Apiservers. Label based cleanup
+ will be implemented as a solution.
+
+One extra artifact that needs to deleted per Stack uninstallation is the token secret created on Host Cluster. This will
+ be achieved by setting owner of that secret as Stack Controller Deployment which also lives same Cluster/Namespace.
+     
+### Packaging Changes
+
+We need to deploy crossplane types and roles (CRDs and RBACs) separately from Crossplane and Stack Manager pods. This 
+ could be realized by introducing some new helm parameters to existing Crossplane helm chart, however, 
+ [`helm 3` does not support templating in CRDs directory](https://github.com/helm/helm/issues/7083) which blocks 
+ conditional installation. 
+ 
+We will introduce following two helm (helm 3) charts:
+
+`crossplane-types`: this is the piece that will be deployed into Tenant Kubernetes. This chart includes CRDs and
```

Comment:
Why and how would helm be involved in priming a tenant cluster?

Could we accomplish this using only the stack-manager, each stack's list of dependencies, and a label on these blessed/core CRDs?  We could copy these host CRDs to tenants during stack install by having the install-job capture the existing CRD from the host, when listed as a dependency, and outputting it to the install-job log.

Maybe the label that permits a CRD to be copied to tenants could look like the following:
```
  labels:
   core.crossplane.io/tenant-safe: True
```

But doing this would necessitate scrubbing of preexisting core.crossplane.io labels from Stack CRDs.

Reply:
This is for installing CRDs currently coming with existing `Crossplane` helm chart. Right now it includes `Stack`, `StackInstall`, `ClusterStackInstall` and core Crossplane types. If we could bundle core crossplane as a stack, then it will only include stack manager types. 
Before installing any stack we need to install Stack Manager (and Crossplane Core). This installation is no longer done into a single Kubernetes, rather we need to split resource and install some to host and some to tenant. `Crossplane-types` helm chart is for the ones to be installed into Tenant.

Reply:
We had a follow-up chat about this resulting in the following table, which helped clarify responsibilities for me:

Kind                  | Name                                | Place
----                  | -----                                | ------
pod                   | crossplane                           | Host (ns: ubc1234-system)
pod                   | stack-manager                        | Host (ns: ubc1234-system)
job                   | (stack installjob)                   | Host (ns: ubc1234-controllers)
pod                   | (stack controller)                   | Host (ns: ubc1234-controllers)

Kind                  | Name                                | Place
----                  | -----                                | ------
crd                   | Stack, SI, CSI                       | Tennant
Stack                 | wordpress                            | Tennant
StackInstall          | wordpress                            | Tennant
crd                   | KubernetesEngine, MysqlInstance, ... | Tennant
crd                   | GKEInstance, CloudSQLInstance, ...   | Tennant
(rbac)                  | (stack controller)                   | Tennant
(rbac)                  | (workspace owner, crossplane-admin)  | Tennant
(rbac)                  | (stack:namespace:1.2.3:admin)        | Tennant
crd                   | WordpressInstance                    | Tennant
WordpressInstance     | wp-instance                          | Tennant
KubernetesApplication | wp-instance                          | Tennant
pod                   | wp-instance (from KubernetesAplication)    | New Cluster